### PR TITLE
Socket reconnects

### PR DESCRIPTION
**Pull request by flxo**
Attempt to reconnect the logd socket upon errors.

### CODE REVIEW

File: src/thread.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+pub fn id() -> usize {
+    unsafe { winapi::um::processthreadsapi::GetCurrentThreadId() as usize }
+}
+
+#[cfg(target_os = "redox")]
+#[inline]
+pub fn id() -> usize {
+    // Each thread has a separate pid on Redox.
+    syscall::getpid().unwrap()
+}
```
Comment (rosds) [inline src/thread.rs:18]: shouldn't `std::thread::current().id()` be equivalent to these?

File: .cargo/runner-aarch64-linux-android [hunk 2]
```diff
@@ -6,6 +6,7 @@ set -x
 BIN=`basename $1`
 
 adb remount
-adb push $1 /data
-adb shell /data/$BIN ${@:2}
-adb shell rm /data/$BIN
+adb push $1 /system/bin
+adb shell /system/bin/$BIN ${@:2}
+adb shell chmod +x /system/bin/$BIN
```
Comment (rosds) [inline .cargo/runner-aarch64-linux-android:11]: is this correct? is it adding exec rights just after executing the binary?

File: src/lib.rs [hunk 3]
```diff
@@ -472,55 +432,22 @@ impl Log for Logger {
         };
 
         let priority: Priority = record.metadata().level().into();
+        let tag = if let Some(ref tag) = self.tag {
+            tag
+        } else {
+            record.module_path().unwrap_or_default()
+        };
```
Comment (rosds) [inline src/lib.rs:439]: nit:
```rust
let tag = self.tag.as_ref().unwrap_or_else(|| record.module_path().unwrap_or_default());
```

File: src/logd.rs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+/// Logd write socket path
+const LOGDW: &str = "/dev/socket/logdw";
+
+lazy_static::lazy_static! {
+    static ref SOCKET: LogdSocket = LogdSocket::connect(Path::new(LOGDW));
+}
+
+/// Logd write socket abstraction. Sends never fail and on each send a reconnect
+/// reconnect attempt is made.
```
Comment (rosds) [inline src/logd.rs:16]: "reconnect" duplicated

File: src/logd.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    socket: parking_lot::RwLock<UnixDatagram>,
+}
+
+impl LogdSocket {
+    /// Construct a new LogdSocket.
+    ///
+    /// # Panics
+    ///
+    /// This function panics when the socket cannot be created.
+    pub fn connect(path: &Path) -> LogdSocket {
```
Comment (rosds) [inline src/logd.rs:27]: shouldn't these methods be private?

File: src/logd.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+/// Logd write socket path
+const LOGDW: &str = "/dev/socket/logdw";
+
+lazy_static::lazy_static! {
+    static ref SOCKET: LogdSocket = LogdSocket::connect(Path::new(LOGDW));
+}
+
+/// Logd write socket abstraction. Sends never fail and on each send a reconnect
+/// reconnect attempt is made.
+struct LogdSocket {
```
Comment (rosds) [inline src/logd.rs:17]: This guy doesn't look to have logic specific to `logd`, maybe just `Socket` or `RetryConnectSocket` is more descriptive?

File: src/logd.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+            let socket = UnixDatagram::unbound()?;
+
+            // Upgrade the read lock and replace the socket if the sent attempt is successful.
+            let mut lock = RwLockUpgradableReadGuard::upgrade(lock);
+            socket.connect(LOGDW)?;
+            socket.send(buffer)?;
+            // Assign the new socket to the lock. In the worst case one or more threads
+            // are opening sockets to logd which are immediately closed.
+            *lock = socket;
+        }
```
Comment (rosds) [inline src/logd.rs:49]: I like the mutex trick to handle the new connection. Just curious about how frequent does this happen and if a single reconnect is enough or if a classic retry loop should be used?

**Review by rosds — APPROVED** [review]
lgtm!

File: src/thread.rs (hunk 1 above)
Comment (flxo) [reply src/thread.rs:18]: I don't remember the reason, but there has been one. Think it was the type of the thread id but I'm unsure.

File: .cargo/runner-aarch64-linux-android (hunk 2 above)
Comment (flxo) [reply .cargo/runner-aarch64-linux-android:11]: good catch. Fixed.

File: src/lib.rs (hunk 3 above)
Comment (flxo) [reply src/lib.rs:439]: Thanks. This needs a `as_deref` because tag is a `Option<String>`.

File: src/logd.rs (hunk 5 above)
Comment (flxo) [reply src/logd.rs:27]: `connect` is called from the `lazy_static` init and therefore must be `pub`.

File: src/logd.rs (hunk 7 above)
Comment (flxo) [reply src/logd.rs:49]: This shouldn't happen at all ;-). A classic retry loop would block a user infinity in the worst case if `logd` is refusing to take a new connection.

File: src/thread.rs (hunk 1 above)
Comment (flxo) [reply src/thread.rs:18]: The reason is:

From `std::thread::ThreadId`:
```
A ThreadId is an opaque object that uniquely [id](https://doc.rust-lang.org/std/thread/struct.Thread.html#method.id)entifies each thread created during the lifetime of a process. [Thread](https://doc.rust-lang.org/std/thread/struct.Thread.html)Ids are guaranteed not to be reused, even when a thread terminates. ThreadIds are under the control of Rust’s standard library and there may not be any relationship between ThreadId and the underlying platform’s notion of a thread identifier – the two concepts cannot, therefore, be used interchangeably. A ThreadId can be retrieved from the id method on a Thread.
```

The logger should use the OS thread id and not a Rust stdlib internal one.

**Pull request merged**